docs/proposals/push-metrics.md

```diff
@@ -0,0 +1,221 @@
+pushed in bulk -- it is up to the format to support a way to name specific
+metrics, namespaces, pods, and containers (for instance, the Prometheus format
+uses metric names and labels for this purpose).
+
+### Metrics Format
+
+The underlying design can support multiple format "backends".  The initial
+backend, detailed here, will be based on the Prometheus format, and will be
+available at `/api/v1/push/prometheus`.
+
+Initially, only the Prometheus text format will be supported. In the future, we
+may also wish to support the promethus protobuf format as well.
+
+The metric name specified on each metric line will be used as the custom metric
+name in Heapster (except prefixed as discussed in the "Authentication"
+section).  The following labels will be used to determine which object a metric
+is associated with:
+
+- For kubernetes-related metrics, the `namespace` label will indicate
+  namespace, the `pod` label will indicate pod, and the `container` label will
+  indicate container.  If only `namespace` is present, then the metric will be
+  considered namespace-level.  If `namespace` and `pod` are present, then the
+  metric will be considered pod-level.  If all three are present, the metric
+  will be considered container-level.
+
+- For kubernetes-related metrics, the `service` label can be used in
+  conjunction with the `namespace` label to indicate a service-level metric.
+  Currently, this will be stored in Heapster as a namespace-level metric
+  prefixed with the service name.
+
+- For non-kubernetes-related metrics, the `node` label will indicate node, and
+  the `container` label will indicate free container.   If only `node` is
+  present, the metric will be considered node-level.  Otherwise, the metric
+  will be considered free-container-level.
+
+If the `node`, `namespace`, `pod`, and `container`, and `service` labels are
+not present in one of the configurations listed above, the metric line is
+invalid and the batch should be rejected.
+
+Any additional labels will be treated the same as labels on existing custom
+metrics (currently multiple custom metrics with the same name, but different
+labels, are ignored, but this seems like an oversight and should probably be
+fixed).
+
+Timestamps will be assigned by using the next Heapster metrics batch timestamp
+after the time at which the metrics are received.  If a timestamp is provided
+as part of the metric line, this may be stored as a separate field for
+posterity, but the "official" timestamp will be that of the assigned batch.
+
+As with normal Prometheus metrics, the `TYPE` line should be used to provide
+the type of the metric.
+
+#### Example
+
+Suppose a producer with the name "http_gatherer" sent the following metrics to
+`/api/v1/push/prometheus`:
+
+```
+# This is a pod-level metric (it might be used for autoscaling)
+# TYPE http_requests_per_second guage
+http_requests_per_minute{namespace="webapp",pod="frontend-server-a-1"} 20
```

Comment:
The very next set of metrics is an example of a service-level metric.  I said above that this is for *cluster components* -- sources producing metrics tied to other pods/services/namespaces/etc.

Currently, we have a method for individual pods to expose information about *themselves* (the cAdvisor-Kubelet-Heapster-Pull chain) This proposal would allow exposing metrics about other entities (pods, services, namespaces, etc) in the cluster.